### PR TITLE
Refactor halide_device_free call injection

### PR DESCRIPTION
**Pull request by dsharletg**
The old code used CodeGen_GPU_Host to add halide_device_free calls and destructors, which doesn't work if the target uses inject_host_dev_buffer_copies, but doesn't use CodeGen_GPU_Host.

The new implementation works as follows:

- When we add a halide_device_malloc call, we also register a destructor (so halide_device_free gets called no matter what happens next, it might just be much later than we want).
- In EarlyFree, for buffers that had halide_device_malloc called, we then inject a halide_device_free call immediately prior to the Free node.

### CODE REVIEW

**Comment by abadams** [comment]
I see the code that adds a device free call prior to the Free node, but I don't see the code that registers a destructor. What am I missing?

**Comment by dsharletg** [comment]
Whoops, I committed that to a different branch on accident (different shell windows != different git contexts...). Should be there now.

**Comment by abadams** [comment]
lgtm

File: src/EarlyFree.cpp [hunk 1]
```diff
@@ -44,6 +47,9 @@ class FindLastUse : public IRVisitor {
         if (call->name == func) {
             last_use = containing_stmt;
         }
+        if (call->name == "halide_device_malloc" && expr_uses_var(call, func + ".buffer")) {
```
Comment (zvookin) [inline src/EarlyFree.cpp]: Is there an issue here with halide_device_malloc for an interface other than the one being considered in this traversal?

Comment (dsharletg) [inline src/EarlyFree.cpp]: I don't think so, that shouldn't matter. Basically, if the buffer ever had any device_malloc on it, regardless of device interface, we should call device_free. It's really doing the same thing as destructor would, just earlier.

**Comment by steven-johnson** [comment]
LGTM. Any objections to merging?

**Comment by dsharletg** [comment]
I'm merging this. Hexagon blatantly leaks device allocations without this change.

**Comment by abadams** [comment]
cuda tests are failing with this in. Something must be wrong:

http://buildbot.halide-lang.org:8010/builders/linux-64-gcc49-380/builds/204/steps/make%20test_correctness_4/logs/stdio

**Comment by abadams** [comment]
I think the problem is that the old code also did a device free on external buffers if they were copied to device in the kernel and then there was an error.

Not sure if that's good or bad. If copy_to_device was a no-op because the buffer was already on the GPU, then it's pretty annoying for Halide to free it on an error.

**Comment by abadams** [comment]
I'll make the Buffer destructor ignore errors in halide_dev_free. Throwing errors in destructors is not good.

**Comment by abadams** [comment]
2/3 of the gpu_object_lifetime tests don't actually test anything, because they never use the target that they configure... when I turn on debug, they fail. I think I need to roll this back.

**Comment by dsharletg** [comment]
That doesn't make sense, when I run those tests explicitly, I get all the debug output and checking that I expect?

**Comment by abadams** [comment]
If you run them explicitly with HL_JIT_TARGET=host-cuda-debug, they fail.
Look to see where target is used in gpu_object_lifetime 2 and 3

**Comment by abadams** [comment]
I pushed fixes to the test in the branch

**Comment by abadams** [comment]
Ok, just needed more test fixes. Corrected version pushed to master.

**Pull request merged**